src/libsync/clientsideencryption.cpp

```diff
@@ -0,0 +1,1168 @@
+            if (file.open(QIODevice::WriteOnly)) {
+                QTextStream s(&file);
+                s << caps;
+            }
+            file.close();
+            qCInfo(lcCse()) << "public key saved, Encrypting Private Key.";
+            encryptPrivateKey(keyPair);
+        }
+        qCInfo(lcCse()) << retCode;
+    });
+    job->start();
+
+free_all:
+    X509_REQ_free(x509_req);
+    BIO_free_all(out);
+}
+
+void ClientSideEncryption::setTokenForFolder(const QByteArray& folderId, const QByteArray& token)
+{
+	_folder2token[folderId] = token;
+}
+
+QByteArray ClientSideEncryption::tokenForFolder(const QByteArray& folderId) const
+{
+	Q_ASSERT(_folder2token.contains(folderId));
+	return _folder2token[folderId];
+}
+
+void ClientSideEncryption::encryptPrivateKey(EVP_PKEY *keyPair)
+{
+    // Write the Private File to a BIO
+    // Retrieve the BIO contents, and encrypt it.
+    // Send the encrypted key to the server.
+    // I have no idea what I'm doing.
+
+    using ucharp = unsigned char *;
+    const char *salt = "$4$YmBjm3hk$Qb74D5IUYwghUmzsMqeNFx5z0/8$";
+    const int saltLen = 40;
+    const int iterationCount = 1024;
+    const int keyStrength = 256;
+    BIO* bio = BIO_new(BIO_s_mem());
+
+    QString passPhrase = WordList::getUnifiedString(WordList::getRandomWords(12));
+    const char* passPhrasePtr = qPrintable(passPhrase);
+    qCInfo(lcCse()) << "Passphrase Generated:";
+    qCInfo(lcCse()) << passPhrase;
+
+    // Extract the Private key from the key pair.
+    PEM_write_bio_PrivateKey(bio, keyPair, NULL, NULL, 0, 0, NULL);
+    char data[80];
+    QString output;
+    int ret = 0;
+    do {
+        ret = BIO_gets(bio, data, 80);
+        output += data;
+        if (output.endsWith("-----END PRIVATE KEY-----")) {
+            output = output.trimmed();
+            break;
+        }
+    } while (ret > 0 );
+
+    qCInfo(lcCse()) << "Private Key Extracted";
+    qCInfo(lcCse()) << output;
+
+    /* Jesus. the OpenSSL docs do not help at all.
+     * This PKCS5_PBKDF2_HMAC_SHA1 call will generate
+     * a new password from the password that was submited.
+     */
+    unsigned char secretKey[keyStrength];
+
+    ret = PKCS5_PBKDF2_HMAC_SHA1(
+        passPhrasePtr,     // const char *password,
+        passPhrase.size(), // int password length,
+        (ucharp) salt,     // const unsigned char *salt,
+        saltLen,      // int saltlen,
+        iterationCount,    // int iterations,
+        keyStrength,       // int keylen,
+        secretKey          // unsigned char *out
+    );
+    qCInfo(lcCse()) << "Return of the PKCS5" << ret;
+    qCInfo(lcCse()) << "Result String" << secretKey;
+
+
+    /**
+      * NOTES: the key + iv have to be base64 encoded in the metadata.
+      */
+
+    //: FIRST TRY A SILLY PHRASE.
+    //: Hardcoed IV, really bad.
+    unsigned char *fakepass = (unsigned char*) "qwertyuiasdfghjkzxcvbnm,qwertyui";
+    unsigned char *iv = (unsigned char *)"0123456789012345";
```

Comment:
Also RAND_bytes

Reply:
For an IV we should likely go with RAND_pseudo_bytes. However, also check the return code https://www.openssl.org/docs/man1.0.2/crypto/RAND_bytes.html